src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs

```diff
@@ -530,11 +646,47 @@ public override ConstructorInfo[] GetConstructors(BindingFlags bindingAttr)
         public override EventInfo[] GetEvents(BindingFlags bindingAttr) => throw new NotSupportedException();
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicEvents | DynamicallyAccessedMemberTypes.NonPublicEvents)]
         public override EventInfo? GetEvent(string name, BindingFlags bindingAttr) => throw new NotSupportedException();
+
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)]
-        public override MethodInfo[] GetMethods(BindingFlags bindingAttr) => throw new NotSupportedException();
+        public override MethodInfo[] GetMethods(BindingFlags bindingAttr)
+        {
+            ThrowIfNotCreated();
+
+            List<MethodInfo> methods = new();
+            foreach (MethodBuilderImpl method in _methodDefinitions)
+            {
+                if (!method.IsConstructor && MatchesTheFilter(method, GetBindingFlags(method), bindingAttr, CallingConventions.Any, method.ParameterTypes))
+                {
+                    methods.Add(method);
+                }
+            }
+
+            return methods.ToArray();
+        }
+
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)]
         protected override MethodInfo? GetMethodImpl(string name, BindingFlags bindingAttr, Binder? binder,
-                CallingConventions callConvention, Type[]? types, ParameterModifier[]? modifiers) => throw new NotSupportedException();
+                CallingConventions callConvention, Type[]? types, ParameterModifier[]? modifiers)
+        {
+            ThrowIfNotCreated();
+            ArgumentNullException.ThrowIfNull(types);
```

Comment:
```suggestion
```
This should not be needed. It was validated by public entrypoint: https://github.com/dotnet/runtime/blob/9080024d1ccd38974f0d297ac6373e6e8d62ec5d/src/libraries/System.Private.CoreLib/src/System/Type.cs#L324-L334

Also, I think `types` should be allowed to be null.

Reply:
Right